### PR TITLE
Allow new disks to be added to running VMs via vmpooler API

### PR DESCRIPTION
**Pull request by sschneid**
Add an additional disk to a running VM via the vmpooler API.

````
$ curl -X POST -H X-AUTH-TOKEN:a9znth9dn01t416hrguu56ze37t790bl \
  --url vmpooler.company.com/api/v1/vm/fq6qlpjlsskycq6/disk/8
````
````json
{
  "ok": true,
  "fq6qlpjlsskycq6": {
    "disk": "+8mb"
  }
}
````

Provisioning and attaching disks can take a moment, but once the task completes it will be reflected in a `GET /v1/vm/<hostname>` query:

````
$ curl --url vmpooler.company.com/api/v1/vm/fq6qlpjlsskycq6
````
````json
{
  "ok": true,
  "fq6qlpjlsskycq6": {
    "template": "debian-7-x86_64",
    "lifetime": 2,
    "running": 0.08,
    "state": "running",
    "disk": [
      "+8mb"
    ],
    "domain": "delivery.puppetlabs.net"
  }
}

### CODE REVIEW

**Comment by sschneid** [comment]
This is deployed on `vmpooler-dev`.

File: lib/vmpooler/pool_manager.rb [hunk 1]
```diff
@@ -299,6 +299,47 @@ def destroy_vm(vm, pool)
...
+        template = $redis.hget('vmpooler__vm__' + vm, 'template')
+        datastore = nil
+
+        $config[:pools].each do |pool|
+          if pool['name'] == template
+            datastore = pool['datastore']
+          end
+        end
+
+        if ((! datastore.nil?) || (! datastore.empty?))
```
Comment (colinPL) [inline lib/vmpooler/pool_manager.rb]: This should be `&&`. It is possible for the `datastore` to not be nil (first check) by being empty.

File: lib/vmpooler/vsphere_helper.rb [hunk 2]
```diff
@@ -12,6 +12,60 @@ def initialize(_vInfo = {})
...
     end
 
+    def add_disk(vmname, size, datastore)
+      begin
+        @connection.serviceInstance.CurrentTime
+      rescue
+        initialize
+      end
+
+      vm = find_vm(vmname) || find_vm_heavy(vmname)[vmname]
```
Comment (colinPL) [inline lib/vmpooler/vsphere_helper.rb]: This is called by `pool_manager` and the `host` or `vm` object already exists from the calling method. Can `vmname` be changed to a VM object instead?

File: lib/vmpooler/vsphere_helper.rb [hunk 3]
```diff
@@ -12,6 +12,60 @@ def initialize(_vInfo = {})
...
+
+      vmdk_datastore = find_datastore(datastore)
+      vmdk_file_name = "#{vmname}/#{vmname}_#{find_vmdks(vmname, datastore).length + 1}.vmdk"
+
+      controller = find_disk_controller(vm)
+
+      vmdk_spec = RbVmomi::VIM::FileBackedVirtualDiskSpec(
+        capacityKb: size.to_i * 1024 * 1024,
```
Comment (colinPL) [inline lib/vmpooler/vsphere_helper.rb]: Will this work? Can a diskspec be created for 0kb disk?

File: lib/vmpooler/vsphere_helper.rb [hunk 4]
```diff
@@ -12,6 +12,60 @@ def initialize(_vInfo = {})
...
+
+      vmdk_backing = RbVmomi::VIM::VirtualDiskFlatVer2BackingInfo(
+        datastore: vmdk_datastore,
+        diskMode: 'persistent',
+        fileName: "[#{vmdk_datastore.name}] #{vmdk_file_name}"
+      )
+
+      device = RbVmomi::VIM::VirtualDisk(
+        backing: vmdk_backing,
+        capacityInKB: size.to_i * 1024 * 1024,
```
Comment (colinPL) [inline lib/vmpooler/vsphere_helper.rb]: Same question as above -- will this work for 0kb? Or is an error raised?

File: lib/vmpooler/vsphere_helper.rb [hunk 5]
```diff
@@ -12,6 +12,60 @@ def initialize(_vInfo = {})
                                          insecure: true
     end
 
+    def add_disk(vmname, size, datastore)
```
Comment (colinPL) [inline lib/vmpooler/vsphere_helper.rb]: Should add validation at the start to ensure that `size` is an integer > `0`.

File: lib/vmpooler/vsphere_helper.rb [hunk 6]
```diff
@@ -12,6 +12,60 @@ def initialize(_vInfo = {})
...
+
+      vmdk_spec = RbVmomi::VIM::FileBackedVirtualDiskSpec(
+        capacityKb: size.to_i * 1024 * 1024,
+        adapterType: 'lsiLogic',
+        diskType: 'thin'
+      )
+
+      vmdk_backing = RbVmomi::VIM::VirtualDiskFlatVer2BackingInfo(
+        datastore: vmdk_datastore,
+        diskMode: 'persistent',
```
Comment (colinPL) [inline lib/vmpooler/vsphere_helper.rb]: Please set this as a constant.

File: lib/vmpooler/vsphere_helper.rb [hunk 7]
```diff
@@ -12,6 +12,60 @@ def initialize(_vInfo = {})
...
+      vm = find_vm(vmname) || find_vm_heavy(vmname)[vmname]
+
+      vmdk_datastore = find_datastore(datastore)
+      vmdk_file_name = "#{vmname}/#{vmname}_#{find_vmdks(vmname, datastore).length + 1}.vmdk"
+
+      controller = find_disk_controller(vm)
+
+      vmdk_spec = RbVmomi::VIM::FileBackedVirtualDiskSpec(
+        capacityKb: size.to_i * 1024 * 1024,
+        adapterType: 'lsiLogic',
```
Comment (colinPL) [inline lib/vmpooler/vsphere_helper.rb]: Please set as a constant.

File: lib/vmpooler/vsphere_helper.rb [hunk 8]
```diff
@@ -12,6 +12,60 @@ def initialize(_vInfo = {})
...
+        diskType: 'thin'
```
Comment (colinPL) [inline lib/vmpooler/vsphere_helper.rb]: Please set as a constant.

File: lib/vmpooler/vsphere_helper.rb [hunk 9]
```diff
@@ -23,6 +77,102 @@ def find_datastore(datastorename)
...
+      end
+
+      vm.config.hardware.device.each do |device|
+        return device if device.deviceInfo.label == deviceName
+      end
+
+      nil
+    end
+
+    def find_disk_controller(vm)
```
Comment (colinPL) [inline lib/vmpooler/vsphere_helper.rb]: Can this look more like `find_device`?

File: lib/vmpooler/vsphere_helper.rb [hunk 10]
```diff
@@ -23,6 +77,102 @@ def find_datastore(datastorename)
...
+      devices.keys.sort.each do |c|
+        next unless controller.key == devices[c]['device'].key
+        used_unit_numbers.push(devices[c]['device'].scsiCtlrUnitNumber)
+        devices[c]['children'].each do |disk|
+          used_unit_numbers.push(disk.unitNumber)
+        end
+      end
+
+      (0..15).each do |scsi_id|
+        if used_unit_numbers.grep(scsi_id).length > 0
```
Comment (colinPL) [inline lib/vmpooler/vsphere_helper.rb]: Change to `<= 0` to remove empty if.

File: lib/vmpooler/pool_manager.rb [hunk 11]
```diff
@@ -299,6 +299,47 @@ def destroy_vm(vm, pool)
...
+        datastore = nil
+
+        $config[:pools].each do |pool|
+          if pool['name'] == template
+            datastore = pool['datastore']
+          end
+        end
+
+        if ((! datastore.nil?) && (! datastore.empty?))
+          $vsphere['disk_manager'].add_disk(vm, disk_size, datastore)
```
Comment (colinPL) [inline lib/vmpooler/pool_manager.rb]: Should this check the return value before proceeding onwards?

File: lib/vmpooler/pool_manager.rb [hunk 12]
```diff
@@ -299,6 +299,47 @@ def destroy_vm(vm, pool)
       end
     end
 
+    def create_vm_disk(vm, disk_size)
+      Thread.new do
+        _create_vm_disk(vm, disk_size)
+      end
+    end
+
+    def _create_vm_disk(vm, disk_size)
```
Comment (colinPL) [inline lib/vmpooler/pool_manager.rb]: Should verify `disk_size` is `> 0`.

File: lib/vmpooler/api/v1.rb [hunk 13]
```diff
@@ -552,6 +556,30 @@ def checkout_vm(template, result)
       JSON.pretty_generate(result)
     end
 
+    post "#{api_prefix}/vm/:hostname/disk/:size/?" do
```
Comment (colinPL) [inline lib/vmpooler/api/v1.rb]: Validate `size`. It is possible for `size` to contain a colon at this time.

File: lib/vmpooler/vsphere_helper.rb (hunk 3 above)
Comment (sschneid) [inline lib/vmpooler/vsphere_helper.rb]: Addressed at https://github.com/sschneid/vmpooler/commit/7d0f7254ae8e538beb92804f91f59547dd2bf53b#diff-c96671f8bdd3f0c734cd3f1fc6730a51R26

File: lib/vmpooler/vsphere_helper.rb (hunk 4 above)
Comment (sschneid) [inline lib/vmpooler/vsphere_helper.rb]: Addressed at https://github.com/sschneid/vmpooler/commit/7d0f7254ae8e538beb92804f91f59547dd2bf53b#diff-c96671f8bdd3f0c734cd3f1fc6730a51R26

File: lib/vmpooler/vsphere_helper.rb (hunk 9 above)
Comment (sschneid) [inline lib/vmpooler/vsphere_helper.rb]: This is done.

File: lib/vmpooler/pool_manager.rb (hunk 12 above)
Comment (sschneid) [inline lib/vmpooler/pool_manager.rb]: Addressed at https://github.com/puppetlabs/vmpooler/pull/147/files#diff-9ca942eee48c471766a71aa4fd1df851R312

File: lib/vmpooler/api/v1.rb (hunk 13 above)
Comment (sschneid) [inline lib/vmpooler/api/v1.rb]: Addressed at https://github.com/puppetlabs/vmpooler/pull/147/files#diff-51d4606238cfb066704749b609c2d728R569

**Comment by sschneid** [comment]
@colinPL I've pushed fixups for all of your comments; ready for re-review.

**Pull request merged**